solutions/platinum/usaco-578.mdx

```diff
@@ -0,0 +1,126 @@
+---
+id: usaco-578
+source: USACO Platinum 2015 December
+title: Counting Haybales
+author: Justin Ji
+---
+
+[Official Analysis](https://usaco.org/index.php?page=viewproblem2&cpid=578)
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+struct LazySegtree {
+	static constexpr array<ll, 2> ID = {0, INT32_MAX};  // default tree value
+	static constexpr int LZ_ID = 0;                     // default lazy value
+	const int sz;
+	vector<array<ll, 2>> t;  // t[v] = {sum, minimum}
+	vector<int> lz;          // lz[v] = lazy add
+
+	/** @return the result of joining two tree nodes */
+	inline array<ll, 2> join(const array<ll, 2> &a, const array<ll, 2> &b) {
+		return {a[0] + b[0], min(a[1], b[1])};
+	}
+
+	LazySegtree(const vector<int> &a)
+	    : sz((int)a.size()), t(4 * sz), lz(4 * sz) {
+		build(1, 0, sz - 1, a);
+	}
+
+	void build(int v, int tl, int tr, const vector<int> &a) {
```

Comment:
why not put directly in constructor

Reply:
like a recursive lambda or smth?

Reply:
can you make this and other functions that aren't directly called private